scrapy/extensions/closespider.py

```diff
@@ -69,3 +82,31 @@ def spider_closed(self, spider):
         task = getattr(self, "task", False)
         if task and task.active():
             task.cancel()
+
+        task_no_item = getattr(self, "task_no_item", False)
+        if task_no_item and task_no_item.running:
+            task_no_item.stop()
+
+    def spider_opened_no_item(self, spider):
+        from twisted.internet import task
+
+        self.task_no_item = task.LoopingCall(self._count_items_produced, spider)
+        self.task_no_item.start(self.timeout_no_item, now=False)
+
+        logger.info(
+            f"Spider will stop when no items are produced after "
+            f"{self.timeout_no_item} seconds."
+        )
+
+    def item_scraped_no_item(self, item, spider):
+        self.items_in_period += 1
+
+    def _count_items_produced(self, spider):
+        if self.items_in_period >= 1:
```

Comment:
I initially thought of exposing this but I think it's better to keep it simple for now. Users can opt to override this method anyway if they need a customized condition of when to close the spider.